### PR TITLE
[CHECKOUTUI-51] Set cart items when when click go forward borwser arr…

### PR DESCRIPTION
**Pull request by Arturo777**
[Ticket](https://vtex-dev.atlassian.net/browse/CHECKOUTUI-51)
[Workspace to test](https://arturovtexapps--mercedesbenzco.myvtex.com/checkout/#/email)
What i'm solving?
when you click on proceed to payment in the checkout you'll see your products listed, but when you go back through the browser arrow, and then go forward either through the browser arrow the items were not appearing. This is because the container element takes some time to be ready before you can add the `v-loaded`class.

Steps to reproduce:
1.- Go to checkout.
2.- Go to proceed payment and see the list of products.
3.- Go back through the browser arrow.
4.- Go forward through the browser arrow.
And see if you can see the list of products.

### CODE REVIEW

File: checkout-ui-custom/src/_js/_v.custom.checkout.ui.js [hunk 1]
```diff
@@ -364,6 +364,19 @@ class checkoutCustom {
...
+    let counter = 0
+    const max = 10
+    const cartItems = setInterval(function() {
+      if ($('.cart-items').length > 0 || counter >= max) {
+        clearInterval(cartItems)
+        $(`.cart-items`).addClass('v-loaded')
+      } else {
+        counter++
+      }
+    }, 500)
```
Comment (arturmagalhaesjr) [inline checkout-ui-custom/src/_js/_v.custom.checkout.ui.js:377]: How did you test it and reach the max = 10 ?
In addition, I'd like to recommend to you test it by changing your network connection to **slow 3g** and check if the "max = 10" is enough.

image

**Comment by lucvysk** [comment]
I didn't follow why not the hashchange...

File: checkout-ui-custom/src/_js/_v.custom.checkout.ui.js (hunk 1 above)
Comment (Arturo777) [reply checkout-ui-custom/src/_js/_v.custom.checkout.ui.js:377]: Hello @arturmagalhaesjr , i decided 10 as max as i tested with the debugger counting manually how many iterations take to be good and a regular number was 4, so 10 i think is good enough, and i just tried with slow 3G and it works either, thank you!

**Comment by Arturo777** [comment]
> 
i'm using the hashchange but i need to wait to the container to add the class in order to the items appear.

File: CHANGELOG.md [hunk 2]
```diff
@@ -7,6 +7,9 @@ and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.
 
 ## [Unreleased]
 
+### Fixed
+- Set cart items when when click go forward borwser arrows.
```
Comment (arturmagalhaesjr) [inline CHANGELOG.md:11]:
``` suggestion
- Set cart items when clicking on the forward browser arrow.
```

**Review by arturmagalhaesjr — CHANGES_REQUESTED** [review]

**Comment by lucvysk** [comment]
> > _I didn't follow why not the hashchange..._
> > i'm using the hashchange but i need to wait to the container to add the class in order to the items appear.

I rather if we could set that up at least in just one place. `setTimeOut` is quite bad and i'm seeing that multiplying across the code again and again. If that is the **last** resort, i think we should have them at least in just one place. Like ```setTimeOut(() => func1(); func2();)```

**Comment by arturmagalhaesjr** [comment]
> > > _I didn't follow why not the hashchange..._
> > > i'm using the hashchange but i need to wait to the container to add the class in order to the items appear.
> 
> I rather if we could set that up at least in just one place. `setTimeOut` is quite bad and i'm seeing that multiplying across the code again and again. If that is the **last** resort, i think we should have them at least in just one place. Like `setTimeOut(() => func1(); func2();)`

I agree with you partially.  Avoiding the setTimeout and setInterval is the best scenario that we could have in any application. However, embracing all functions inside the same setTimeout/setInterval is not a good idea once we would have a high control on the exceptions once we put all calls on the same "stack";

Javascript event loop is prepared to handle many asynchronous functions as needed, what it could drop the performance is the amount of js processing, example: an eternal setInterval without clearing it, or a hard loop or massive dom changing;

If you are getting some performance issues (I think that is not this case), you should run some performance analysis by chrome performance analysis, lighthouse or GTMetrix;

Also, what we could look at here if we find some javascript blockers on those setTimeouts;

**Comment by arturmagalhaesjr** [comment]
@Arturo777 
Even though we have the tickets, it's important to describe the problem and what you are solving on the PR description once someone can search on the github and find it easily.
Besides that, the steps to reproduce it.

File: checkout-ui-custom/src/_js/_v.custom.checkout.ui.js [hunk 3]
```diff
@@ -47,28 +47,22 @@ class checkoutCustom {
...
 
   checkForFreightSimulation(func) {
-    // Wait until it have the vtex runtime to call the functions
-    const checkRuntime = setInterval(function() {
+    const _this = this
+
+    const isRenderRuntime =
+      window.vtex !== undefined && window.vtex.renderRuntime !== undefined
+
+    _this.awaitForElement(isRenderRuntime, 20, () => {
```
Comment (arturmagalhaesjr) [inline checkout-ui-custom/src/_js/_v.custom.checkout.ui.js:55]: I would create a const in order to make the code clearer:
Example
```ts
const MAX_SET_INTERNAVAL_LOOP = 20
```

```suggestion
   _this.awaitForElement(isRenderRuntime, MAX_SET_INTERNAVAL_LOOP, () => {
```

**Review by arturmagalhaesjr — CHANGES_REQUESTED** [review]

File: checkout-ui-custom/src/_js/_v.custom.checkout.ui.js (hunk 3 above)
Comment (Arturo777) [reply checkout-ui-custom/src/_js/_v.custom.checkout.ui.js:55]: But each element would need a different max, or do you mean, set a const before each call `awaitForElement` function?

Comment (arturmagalhaesjr) [reply checkout-ui-custom/src/_js/_v.custom.checkout.ui.js:55]: yes, absolutely, only to turn it clearer instead of using literals as parameters.

**Review by arturmagalhaesjr — CHANGES_REQUESTED** [review]
yes, absolutely, only to turn it clearer instead of using literals as parameters.

**Review by arturmagalhaesjr — APPROVED** [review]
LGTM

**Pull request merged**